### DISCUSSION TITLE
Vitest is crawling through tests files and running their imports before running setupFiles' content?

### DISCUSSION THREAD

**Discussion by alexandre-ducarne** [vitest-dev/vitest#4456] [Q&A] [answered]
I just moved from Jest to Vitest and it looks like Vitest is crawling through my tests, thus loading all the imports, this breaks the expected flow as some of the imports are accessing variables that are not set and should have been.
Here is an example of what I mean.

```javascript

//vitest.setupFiles.ts

beforeAll(async()=>{
   await Config.fetchSecretsFromGoogleCloudSecrets(); // this will populate the Config abstract class
   console.log(Config.getServerConfig().technicalUserEmail) // "technical@test.com"
})

```

However, what I notice is that if I create a test like so : 

```javascript
// add.test.ts

import { Config } from "../src/config/config";

test("addition", () => {
  const cfg = Config.getServerConfig();
  console.log(cfg.technicalUserEmail) // "technical@test.com"
  expect(1 + 1).toEqual(2);
});

```
my Config object is indeed set and the property is accessible.

However, if I move the 2 first instuctions up, I notice that Vitest will execute everything in my test file except the content of the test function : 

```javascript
// add.test.ts

import { Config } from "../src/config/config";

const email = Config.getServerConfig().technicalUserEmail // ERROR, cannot read properties of undefined (reading 'technicalUserEmail')

test("addition", async () => {
  expect(1 + 1).toEqual(2);
});

```

While the easy fix for this example is to put everything in the test function, it becomes unmanageable when it's the code loaded by an import in the test file that is accessing this kind of variable, ie : 

```javascript
// myModel.ts
export const myModel = mongoose.model("myModel", modelSchema)
// when mongoose.model is called, it requires an active connection to mongoose so that it creates the tables or indexes for instance, that mongoose.connect() that is done in the app needs the Secrets that are fetched like above
```

```javascript
// add.test.ts

import { myModel } from "../src/models/myModel"; // Error, cannot read properties of undefined

test("addition", async () => {
  expect(1 + 1).toEqual(2);
});

```

A solution I found would be to actually do the import in the test function but it's really not pratical and I don't want to do it for the 300+ tests we have : 

```javascript
// add.test.ts

test("addition", async () => {
  const { myModel } = await import ("../src/models/myModel"); //works but requires a lot of refactor
  expect(1 + 1).toEqual(2);
});

```

Do you know what I could do ?

**Comment by sheremet-va** [comment, marked as answer]
First, the setup file is called. Then all tests are collected - to do that, file is imported and executed. Then all `beforeAll` hooks are running for the top level test context, then all test handlers (and their `beforeEach`) in the top level context are running. After that `afterAll` is called. Then the same happens for suites inside `describe`. 

If you need to do something in the setup file before tests are imported, you can call it in the top level of setup file:

```ts
// setupFile.ts
import { Config } from './config.js'
await Config.fetchSecretsFromGoogleCloudSecrets(); // this will populate the Config abstract class
```
